src/mpi/init/mpir_init.c

```diff
@@ -352,3 +356,26 @@ int MPIR_Finalize_impl(void)
 {
     return MPII_Finalize(NULL);
 }
+
+static void qmpi_teardown(void)
+{
+    int counter = 0;
+
+    MPL_free(MPIR_QMPI_pointers);
+    MPL_free(MPIR_QMPI_storage);
+    if (MPIR_CVAR_QMPI_TOOL_LIST != NULL) {
+        /* We allocate one extra tool name so we don't have to shift to avoid the MPICH functions */
+        MPL_free(MPIR_QMPI_tool_names[counter++]);
+        MPL_free(MPIR_QMPI_tool_names[counter++]);
+        size_t len = strlen(MPIR_CVAR_QMPI_TOOL_LIST);
+        for (int i = 0; i <= len; i++) {
+            if (MPIR_CVAR_QMPI_TOOL_LIST[i] == ':') {
+                MPL_free(MPIR_QMPI_tool_names[counter++]);
+            }
+        }
+    }
+    MPL_free(MPIR_QMPI_first_fn_ptrs);
+    MPL_free(MPIR_QMPI_first_tool_ids);
+    MPL_free(MPIR_QMPI_tool_names);
+    MPL_free(MPIR_QMPI_tool_init_callbacks);
+}
```

Comment:
Is it practical to impose a maximum number of tools and use static memory to avoid these cleanup? The tools in priciple lives outside the MPI_Init/Finalize, right? With session, all scenarios may be possible, so it can be quite complicated.

Reply:
It's possible and certainly practical, though it also is probably against the spirit of what QMPI is supposed to do, which is to allow an unlimited number of tools.